airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobCleaner.java

```diff
@@ -0,0 +1,165 @@
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void deleteOldFiles() throws IOException {
+    final Date oldestAllowed = getDateFromDaysAgo(maxWorkspaceRetentionDays);
+
+    Files.walk(workspaceRoot)
+        .map(Path::toFile)
+        .filter(f -> new AgeFileFilter(oldestAllowed).accept(f))
+        .forEach(file -> {
+          LOGGER.info("Deleting old file: " + file.toString());
+          FileUtils.deleteQuietly(file);
+
+          File parentDir = file.getParentFile();
+          if (parentDir.isDirectory() && parentDir.listFiles().length == 0) {
+            FileUtils.deleteQuietly(parentDir);
+          }
+        });
+  }
+
+  private void deleteOnSize() throws IOException {
+    Set<String> nonTerminalJobIds = new HashSet<>();
+    Sets.SetView<JobStatus> nonTerminalStatuses = Sets.difference(Set.of(JobStatus.values()), JobStatus.TERMINAL_STATUSES);
+
+    for (JobStatus nonTerminalStatus : nonTerminalStatuses) {
+      Set<String> jobIds = jobPersistence.listJobsWithStatus(nonTerminalStatus)
+          .stream()
+          .map(Job::getId)
+          .map(String::valueOf)
+          .collect(Collectors.toSet());
+
+      nonTerminalJobIds.addAll(jobIds);
+    }
+
+    final Date youngestAllowed = getDateFromDaysAgo(minWorkspaceRetentionDays);
+
+    final long workspaceBytes = FileUtils.sizeOfDirectory(workspaceRoot.toFile());
+    AtomicLong deletedBytes = new AtomicLong(0);
+    final AgeFileFilter ageFilter = new AgeFileFilter(youngestAllowed);
+    Files.walk(workspaceRoot)
+        .map(Path::toFile)
+        .filter(f -> {
+          String[] splits = f.toPath().toString().replace(workspaceRoot.toString(), "").split("/");
+          if (splits.length >= 2) {
+            return !nonTerminalJobIds.contains(splits[1]);
+          } else {
+            return true;
+          }
+        })
+        .filter(ageFilter::accept)
+        .sorted((o1, o2) -> {
+          FileTime ft1 = getFileTime(o1);
+          FileTime ft2 = getFileTime(o2);
+          return ft1.compareTo(ft2);
+        })
+        .forEach(fileToDelete -> {
+          if (workspaceBytes - deletedBytes.get() > maxWorkspaceSizeMb * 1024 * 1024) {
+            long sizeToDelete = fileToDelete.length();
+            deletedBytes.addAndGet(sizeToDelete);
+            LOGGER.info("Deleting: " + fileToDelete.toString());
+            FileUtils.deleteQuietly(fileToDelete);
+
+            File parentDir = fileToDelete.getParentFile();
+            if (parentDir.isDirectory() && parentDir.listFiles().length == 0) {
+              FileUtils.deleteQuietly(parentDir);
+            }
+          }
+        });
+  }
+
+  protected static Date getDateFromDaysAgo(long daysAgo) {
+    return Date.from(LocalDateTime.now().minusDays(daysAgo).toInstant(OffsetDateTime.now().getOffset()));
+  }
+
+  private static FileTime getFileTime(File file) {
+    try {
+      return Files.readAttributes(file.toPath(), BasicFileAttributes.class).creationTime();
```

Comment:
I dno't think this will bite us here, but not all FSs know when a file was created e.g: linux doesn't know

Reply:
This method is pretty reasonable:
```
     * <p> If the file system implementation does not support a time stamp
     * to indicate the time when the file was created then this method returns
     * an implementation specific default value, typically the {@link
     * #lastModifiedTime() last-modified-time} or a {@code FileTime}
     * representing the epoch (1970-01-01T00:00:00Z).
     ```